### PR TITLE
Refactor to add advanced job submission

### PR DESCRIPTION
**Pull request by jkwmoore**
Still in draft and I cannot remember much of what I got up to.

### CODE REVIEW

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+  $ sbatch --array=0-31
+
+  # Submit a job array with index values of 1, 3, 5 and 7
+  $ sbatch --array=1,3,5,7
+
+  # Submit a job array with index values between 1 and 7
+  # with a step size of 2 (i.e. 1, 3, 5 and 7)
+  $ sbatch --array=1-7:2
+
+Job arrays will have two additional environment variable set. ``SLURM_ARRAY_JOB_ID``
```
Comment (Robadob) [inline hpc/scheduler/advanced_job_submission_and_control.rst:239]: >two

But you listed five

Comment (Robadob) [reply hpc/scheduler/advanced_job_submission_and_control.rst:239]: Rest of the bessemer job arrays section reads well to me.

Only tertiary info I can think of that would be a convenient addition is a short example for people who want to map the 1D index into 2D(+) indices. I anticipate that's a common requirement for people.

**Comment by willfurnass** [comment]
@Robadob has just reminded us elsewhere that this is outstanding!

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+.. code-block:: console
+
+  $ sbatch --array=1-3
+  Submitted batch job 36
+
+Will generate a job array containing three jobs with the environment variables set as follows:
+
+.. code-block:: shell
+
+  SLURM_JOB_ID=36
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:258]: Maybe have a SLURM_JOB_ID more like those often seen, such as 38240.

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+
+Timed start jobs in SGE scheduler are requested with the ``-a`` argument in the following formats:
+
+.. code-block:: shell
+
+  qsub -a 12241200 job1.sh # Dec 24th at 12:00.00
+  qsub -a 202312241200 job2.sh # Dec 24th 2023 at 12:00.00
+  qsub -a 202312241200.30 job2.sh # Dec 24th 2023 at 12:00.30
+
+The scheduler will submit these jobs immediately but wait until the elected time has passed to start the jobs running.
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:464]:
```suggestion
The scheduler will immediately submit these jobs but wait until the elected date/time has passed before starting the execution of the jobs.
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+Timed start jobs using the Slurm scheduler are requested with the ``--begin`` argument in the following formats:
+
+.. code-block:: shell
+
+  sbatch --begin=16:00 job.sh
+  sbatch --begin=now+60 job.sh #(seconds by default)
+  sbatch --begin=now+1hour job.sh
+  sbatch --begin=2023-06-30T12:34:00 job.sh
+
+The scheduler will submit these jobs immediately but wait until the elected time has passed to start the jobs running.
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:492]:
```suggestion
The scheduler will immediately submit these jobs but wait until the elected date/time has passed before starting the execution of the jobs.
```

**Review by Carldkennedy — APPROVED** [review]
LGTM. Left a few suggestions

File: hpc/scheduler/advanced_job_submission_and_control.rst (hunk 2 above)
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:258]:
```suggestion
  SLURM_JOB_ID=39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst (hunk 3 above)
Comment (jkwmoore) [reply hpc/scheduler/advanced_job_submission_and_control.rst:464]:
```suggestion
The scheduler will immediately submit these jobs but wait until the elected date/time has passed before starting these jobs.
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+.. code-block:: shell
+
+  SLURM_JOB_ID=39319
+  SLURM_ARRAY_JOB_ID=36
+  SLURM_ARRAY_TASK_ID=1
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=37
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:265]:
```suggestion
  SLURM_JOB_ID=39320
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+  SLURM_ARRAY_JOB_ID=36
+  SLURM_ARRAY_TASK_ID=2
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=38
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:272]:
```suggestion
  SLURM_JOB_ID=39321
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+
+  $ sbatch --array=1-3
+  Submitted batch job 36
+
+Will generate a job array containing three jobs with the environment variables set as follows:
+
+.. code-block:: shell
+
+  SLURM_JOB_ID=39319
+  SLURM_ARRAY_JOB_ID=36
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:259]:
```suggestion
  SLURM_ARRAY_JOB_ID=39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+  SLURM_ARRAY_TASK_ID=1
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=37
+  SLURM_ARRAY_JOB_ID=36
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:266]:
```suggestion
  SLURM_ARRAY_JOB_ID=39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+  SLURM_ARRAY_TASK_ID=2
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=38
+  SLURM_ARRAY_JOB_ID=36
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:273]:
```suggestion
  SLURM_ARRAY_JOB_ID=39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+  SLURM_ARRAY_TASK_ID=3
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+All SLURM commands and APIs recognize the ``SLURM_JOB_ID`` value. Most commands also recognize the ``SLURM_ARRAY_JOB_ID`` 
+plus ``SLURM_ARRAY_TASK_ID`` values separated by an underscore as identifying an element of a job array. Using the 
+example above, "37" or "36_2" would be equivalent ways to identify the second array element of job 36.
```
Comment (Carldkennedy) [inline hpc/scheduler/advanced_job_submission_and_control.rst:281]:
```suggestion
example above, "39320" or "39319_2" would be equivalent ways to identify the second array element of job 39319. 
```

**Review by Carldkennedy — COMMENTED** [review]
Added a few suggestions based on changing of initial SLURM_JOB_ID

File: hpc/scheduler/advanced_job_submission_and_control.rst (hunk 7 above)
Comment (jkwmoore) [reply hpc/scheduler/advanced_job_submission_and_control.rst:259]: Have you checked what values you get under `SLURM_ARRAY_JOB_ID` and this change matches the format?

File: hpc/scheduler/advanced_job_submission_and_control.rst (hunk 8 above)
Comment (jkwmoore) [reply hpc/scheduler/advanced_job_submission_and_control.rst:266]: As above, have you checked what values you get under SLURM_ARRAY_JOB_ID and this change matches the format?

If it does would it not be 39320?

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+index value. ``SLURM_ARRAY_TASK_COUNT`` will be set to the number of tasks in the job array. 
+``SLURM_ARRAY_TASK_MAX`` will be set to the highest job array index value. 
+``SLURM_ARRAY_TASK_MIN`` will be set to the lowest job array index value. 
+
+For example a job submission of this sort:
+
+.. code-block:: console
+
+  $ sbatch --array=1-3
+  Submitted batch job 36
```
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:252]:
```suggestion
  Submitted batch job 39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+
+  SLURM_JOB_ID=39319
+  SLURM_ARRAY_JOB_ID=39319
+  SLURM_ARRAY_TASK_ID=1
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=39320
+  SLURM_ARRAY_JOB_ID=39319
```
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:266]:
```suggestion
  SLURM_ARRAY_JOB_ID=39320
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+
+  SLURM_JOB_ID=39320
+  SLURM_ARRAY_JOB_ID=39320
+  SLURM_ARRAY_TASK_ID=2
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+  SLURM_JOB_ID=39321
+  SLURM_ARRAY_JOB_ID=39321
```
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:273]:
```suggestion
  SLURM_ARRAY_JOB_ID=39319
```

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+  SLURM_JOB_ID=39321
+  SLURM_ARRAY_JOB_ID=39319
+  SLURM_ARRAY_TASK_ID=3
+  SLURM_ARRAY_TASK_COUNT=3
+  SLURM_ARRAY_TASK_MAX=3
+  SLURM_ARRAY_TASK_MIN=1
+
+All SLURM commands and APIs recognize the ``SLURM_JOB_ID`` value. Most commands also recognize the ``SLURM_ARRAY_JOB_ID`` 
+plus ``SLURM_ARRAY_TASK_ID`` values separated by an underscore as identifying an element of a job array. Using the 
+example above, "39320" or "39319_2" would be equivalent ways to identify the second array element of job 39319.
```
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:281]: I think this is at least partially unclear as we mention 3 different variables and say you can use it to identify the job, but then don't explicitly state you supply it as the `$JOBID` where `sacct -j $JOBID`

File: hpc/scheduler/advanced_job_submission_and_control.rst [hunk 15]
```diff
@@ -257,28 +257,61 @@ Will generate a job array containing three jobs with the environment variables s
...
-  SLURM_ARRAY_TASK_ID=3
+  SLURM_ARRAY_TASK_ID=2
   SLURM_ARRAY_TASK_COUNT=3
   SLURM_ARRAY_TASK_MAX=3
   SLURM_ARRAY_TASK_MIN=1
 
 All SLURM commands and APIs recognize the ``SLURM_JOB_ID`` value. Most commands also recognize the ``SLURM_ARRAY_JOB_ID`` 
 plus ``SLURM_ARRAY_TASK_ID`` values separated by an underscore as identifying an element of a job array. Using the 
-example above, "39320" or "39319_2" would be equivalent ways to identify the second array element of job 39319. 
+example above, "39320" or "39319_1" would be equivalent ways to identify the second array element of job 39319 as
```
Comment (jkwmoore) [inline hpc/scheduler/advanced_job_submission_and_control.rst:281]: This is surprisingly accurate given the SchedMD documentation, so sadly we probably now need to explain:

- The parent job runs the final task (3 in this context) 
- The children jobs start from task minimum array value increasing to the maximum value excluding the final task.

And also submit an issue or bug report to SchedMD asking for clarification / correction on https://slurm.schedmd.com/job_array.html#env_vars

Noting SLURM versions are:

Bessemer - 22.05.6
Stanage - 22.05.4

Good catch!

Debug output from Bessemer:


```
SLURM_JOB_ID=2972603 
SLURM_ARRAY_JOB_ID=2972603 
SLURM_ARRAY_TASK_ID=3 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1

SLURM_JOB_ID=2972604 
SLURM_ARRAY_JOB_ID=2972603 
SLURM_ARRAY_TASK_ID=1 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1

SLURM_JOB_ID=2972605 
SLURM_ARRAY_JOB_ID=2972603 
SLURM_ARRAY_TASK_ID=2 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1
```


Debug output from Stanage:

```
SLURM_JOB_ID=39874 
SLURM_ARRAY_JOB_ID=39874 
SLURM_ARRAY_TASK_ID=3 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1
 
SLURM_JOB_ID=39875 
SLURM_ARRAY_JOB_ID=39874 
SLURM_ARRAY_TASK_ID=1 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1
 
SLURM_JOB_ID=39876 
SLURM_ARRAY_JOB_ID=39874 
SLURM_ARRAY_TASK_ID=2 
SLURM_ARRAY_TASK_MAX=3 
SLURM_ARRAY_TASK_MIN=1
```

Comment (Carldkennedy) [reply hpc/scheduler/advanced_job_submission_and_control.rst:281]: Thanks @jkwmoore https://bugs.schedmd.com/show_bug.cgi?id=16962#c0

**Review by nikishe — APPROVED** [review]
LGTM post chisolms and the job ID changes.

**Pull request merged**